components/script/script_thread.rs

```diff
@@ -326,11 +326,58 @@ impl OpaqueSender<CommonScriptMsg> for Sender<MainThreadScriptMsg> {
     }
 }
 
+/// The state that a document managed by this script thread could be in.
+#[must_root]
+enum DocumentState {
+    /// A document that we are currently keeping.
+    /// The boolean flag indicates whether the script thread is keeping the document alive.
+    /// When documents are active, the flag is true, but if they become inactive we
+    /// set it to false, at which point the document can be garbage collected.
+    Kept(bool, JS<Document>),
+    Discarded(ServoUrl),
+}
+
+unsafe impl JSTraceable for DocumentState {
+    unsafe fn trace(&self, trc: *mut JSTracer) {
+        // Documents in the state `Kept(true, doc)` are kept alive by the script thread,
+        // but documents in the state `Kept(false, doc)` are weak references, that are not
+        // kept alive by the script thread, but may be kept alive by references from other pipelines,
+        // for example by using `iframe.contentDocument`.
+        if let DocumentState::Kept(true, ref doc) = *self {
+            doc.trace(trc);
+        }
+    }
+}
+
+impl DocumentState {
+    fn new(document: &Document) -> DocumentState {
+        DocumentState::Kept(true, JS::from_ref(document))
+    }
+
+    // This function is safe as long as Document implements Drop
+    // to set the state to be discarded.
```

Comment:
I can add a compile test that Document implements Drop, I can't think of a way to actually test the behaviour, since building a Document object is non-trivial.